From f3a7069c1bf281c8b75f66ba277dbbf313bec201 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Mon, 29 Feb 2016 16:33:38 +0200
Subject: [PATCH 0925/1240] tools: a3700: Separate clock setup from DDR in TIM

- Remove clocks configuration from DDR init TIM
  descriptor submodule
- Add TIM submodules for clock presets - 600/600,
  800/800 and 1000/800
- Add support for clock preset submodules into TIM
  builder script
- Add support for clock preset extraction from DTS to
  the common Makefile.

Change-Id: Icb79d8f9082244a46227b38a89c41aed7d275de6
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/27945
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Hanna Hawa <hannah@marvell.com>
---
 Makefile                      | 11 ++++++----
 scripts/buildtim.sh           | 50 ++++++++++++++++++++++++++++++++-----------
 tools/wtp/clocks-1000-800.txt | 50 +++++++++++++++++++++++++++++++++++++++++++
 tools/wtp/clocks-600-600.txt  | 50 +++++++++++++++++++++++++++++++++++++++++++
 tools/wtp/clocks-800-800.txt  | 50 +++++++++++++++++++++++++++++++++++++++++++
 tools/wtp/ddr-800.txt         | 50 -------------------------------------------
 tools/wtp/dll_tune.txt        |  2 ++
 7 files changed, 196 insertions(+), 67 deletions(-)
 create mode 100644 tools/wtp/clocks-1000-800.txt
 create mode 100644 tools/wtp/clocks-600-600.txt
 create mode 100644 tools/wtp/clocks-800-800.txt

diff --git a/Makefile b/Makefile
index 6da47b7..ce55233 100644
--- a/Makefile
+++ b/Makefile
@@ -850,8 +850,9 @@ BIN2PHEX	:= $(srctree)/scripts/bin2phex.pl
 TIM2PHEX	:= $(srctree)/scripts/tim2phex.pl
 TIM2IMG		:= $(srctree)/scripts/tim2img.pl
 TIMBUILD	:= $(srctree)/scripts/buildtim.sh
-
-DDRCFG		:= $(srctree)/tools/wtp/ddr-600.txt
+CLOCKSPATH	:= $(srctree)/tools/wtp
+CLOCKSPRESET	:= $(shell find $(srctree) -name $(CONFIG_DEFAULT_DEVICE_TREE).dts | \
+			xargs grep preset | sed -n 's/.*<\([^ ]*\)>;/\1/p')
 
 ifeq ($(CONFIG_MVEBU_SPI_BOOT),y)
 BOOTDEV		:= SPINOR
@@ -882,8 +883,10 @@ TIM2IMGARGS	:= -i $(DOIMAGE_CFG)
 endif #CONFIG_MVEBU_SECURE_BOOT
 
 TIM_IMAGE	:= $(shell grep "Image Filename:" -m 1 $(DOIMAGE_CFG) | cut -c 17-)
-TIMBLDARGS	:= $(SECURE) $(BOOTDEV) $(IMAGESPATH) $(DDRCFG) $(PARTNUM) $(DOIMAGE_CFG) $(TIMNCFG) $(TIMNSIG)
-TIMBLDUARTARGS	:= $(SECURE) UART $(IMAGESPATH) $(DDRCFG) 0 $(DOIMAGE_CFG) $(TIMNCFG) $(TIMNSIG)
+TIMBLDARGS	:= $(SECURE) $(BOOTDEV) $(IMAGESPATH) $(CLOCKSPATH) $(CLOCKSPRESET) \
+			$(PARTNUM) $(DOIMAGE_CFG) $(TIMNCFG) $(TIMNSIG)
+TIMBLDUARTARGS	:= $(SECURE) UART $(IMAGESPATH) $(CLOCKSPATH) $(CLOCKSPRESET) \
+			0 $(DOIMAGE_CFG) $(TIMNCFG) $(TIMNSIG)
 UARTIMGARCH	:= $(srctree)/uart-images
 
 DOIMAGE_FLAGS := -r $(DOIMAGE_CFG)
diff --git a/scripts/buildtim.sh b/scripts/buildtim.sh
index b0dae55..d7a2f5b 100755
--- a/scripts/buildtim.sh
+++ b/scripts/buildtim.sh
@@ -5,18 +5,20 @@
 # $1 - trusted/non trusted (1/0)
 # $2 - Boot device (SPINOR/SPINAND/EMMCNORM/EMMCALT/SATA/UART)
 # $3 - Path to image text files
-# $4 - DDR init data text file
-# $5 - Partition number
-# $6 - Output TIM/NTIM file name
-# $7 - Output TIMN file name (valid for trusted mode only)
-# $8 - TIMN CSK sign key file name (valid for trusted mode only)
+# $4 - Clocks and DDR init data path
+# $5 - Clocks preset
+# $6 - Partition number
+# $7 - Output TIM/NTIM file name
+# $8 - Output TIMN file name (valid for trusted mode only)
+# $9 - TIMN CSK sign key file name (valid for trusted mode only)
 #
 
 DATE=`date +%d%m%Y`
 IMGPATH=$3
-DDRFILE=$4
-BOOTPART=$5
-OUTFILE=$6
+CLOCKSPATH=$4
+PRESET=$5
+BOOTPART=$6
+OUTFILE=$7
 
 
 # All file names extention
@@ -30,7 +32,7 @@ RSRVDPREF="rsrvd"
 RSRVDLEN=`wc -l < $IMGPATH/$RSRVDPREF.$FILEEXT`
 # DLL tuning - same for all DDR frequencies
 # Located in the same folder as DDR init file
-DLLTUNFILE=$(dirname "$DDRFILE")/"dll_tune."$FILEEXT
+DLLTUNFILE=$CLOCKSPATH/dll_tune.$FILEEXT
 
 # TIM/NTIM image definition file name prefix
 TIMPREF="tim"
@@ -40,8 +42,8 @@ CSKPREF="csk"
 KAKPREF="kak"
 
 # Below values used only by TIMN
-TIMNOUTFILE=$7
-SIGNFILE=$8
+TIMNOUTFILE=$8
+SIGNFILE=$9
 # TIMN image definition file name prefix
 TIMNPREF="timn"
 # Reserved area definition for TIMN - file name prefix
@@ -51,11 +53,12 @@ usage () {
 	echo ""
 	echo "$0 - script for creating TIM/NTIM/TIMN image descriptors"
 	echo ""
-	echo "$0 <trusted> <boot_device> <files_path> <DDR> <output> [timN_out] [timN_key]"
+	echo "$0 <trusted> <boot_device> <files_path> <clocks_path> <clocks_preset> <output> [timN_out] [timN_key]"
 	echo " <trusted>     - trusted/non trusted (supported values 0 and 1)"
 	echo " <boot_device> - Boot device (Supported values SPINOR/SPINAND/EMMCNORM/EMMCALT/SATA/UART)"
 	echo " <files_path>  - Path to image and keys descriptors text files"
-	echo " <DDR>         - DDR initialization sequence text file"
+	echo " <clocks_path> - Path to clocks and DDR initialization files"
+	echo " <clocks_preset> - Name of clocks preset to use - see \"freq\" parameter in DTS file for details"
 	echo " <output>      - Output TIM/NTIM file name"
 	echo " [timN_out]    - Output TIMN file name (required for trusted boot only)"
 	echo " [timN_key]    - TIMN CSK sign key file name (required for trusted boot only)"
@@ -107,6 +110,26 @@ UART)
 	usage
 esac
 
+case "$PRESET" in
+PRESET_CPU_600_DDR_600)
+	CLOCKSFILE=$CLOCKSPATH/clocks-600-600.$FILEEXT
+	DDRFILE=$CLOCKSPATH/ddr-600.$FILEEXT
+	;;
+PRESET_CPU_800_DDR_800)
+	CLOCKSFILE=$CLOCKSPATH/clocks-800-800.$FILEEXT
+	DDRFILE=$CLOCKSPATH/ddr-800.$FILEEXT
+	;;
+PRESET_CPU_1000_DDR_800)
+	CLOCKSFILE=$CLOCKSPATH/clocks-1000-800.$FILEEXT
+	DDRFILE=$CLOCKSPATH/ddr-800.$FILEEXT
+	;;
+*)
+	echo "Unsupported clock preset $PRESET!"
+	usage
+esac
+
+
+
 if [ ! -e "$DDRFILE" ]; then
 	echo "Cannot find DDR init file!"
 	usage
@@ -210,6 +233,7 @@ else
 	echo "DDR_INIT_ENABLE: 0x00000001" >> $OUTFILE
 	echo "End Operations:" >> $OUTFILE
 	echo "Instructions:" >> $OUTFILE
+	cat $CLOCKSFILE >> $OUTFILE
 	cat $DDRFILE >> $OUTFILE
 	cat $DLLTUNFILE >> $OUTFILE
 	echo "End Instructions:" >> $OUTFILE
diff --git a/tools/wtp/clocks-1000-800.txt b/tools/wtp/clocks-1000-800.txt
new file mode 100644
index 0000000..3ef7e86
--- /dev/null
+++ b/tools/wtp/clocks-1000-800.txt
@@ -0,0 +1,50 @@
+; Switch all clocks to REFCLOCK
+WRITE: 0XC0013010 0x00000000
+WRITE: 0XC0018010 0x00000000
+; TBG-A: SE vco_div 0x1,
+; DIFF vco_div 0x1, vco_range 0xa
+; tbg_N 0x30 KVCO = 1600 MHz
+WRITE: 0xC0013204 0x00F00091
+WRITE: 0xC0013204 0x00F000C1
+WRITE: 0xC0013220 0x08030803
+WRITE: 0xC0013208 0x94011400
+WRITE: 0xC0013230 0x00040002
+WRITE: 0xC0013208 0x94011400
+WRITE: 0xC001320C 0x56E656E5
+WRITE: 0xC0013210 0x014B014A
+WRITE: 0xC001320C 0x56E656E5
+WRITE: 0xC0013204 0x00F000C0
+WRITE: 0xC0013208 0x94011400
+WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
+DELAY: 0x00000100
+; TBG-B: SE vco_div 0x1,
+; DIFF vco_div 0x2, vco_range 0xb
+; tbg_N 0x3c KVCO = 2000 MHz
+WRITE: 0xC0013204 0x00F100C0
+WRITE: 0xC0013204 0x00F100C0
+WRITE: 0xC0013220 0x08030803
+WRITE: 0xC0013208 0x14019400
+WRITE: 0xC0013230 0x00040002
+WRITE: 0xC0013208 0x14019400
+WRITE: 0xC001320C 0x56E656E5
+WRITE: 0xC0013210 0x014B014A
+WRITE: 0xC001320C 0x56E656E5
+WRITE: 0xC0013204 0x00F000C0
+WRITE: 0xC0013208 0x14019400
+WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
+DELAY: 0x00000100
+; Set clocks to 1000/800 preset
+WRITE: 0xC0013014 0x00000000
+WRITE: 0xC0013004 0x1296202C
+WRITE: 0xC0013008 0x21061AA9
+WRITE: 0xC001300C 0x20543082
+WRITE: 0xC0013000 0x03CFCCF2
+WRITE: 0xC0018014 0x00180000
+WRITE: 0xC0018004 0x02515508
+WRITE: 0xC0018008 0x00307880
+WRITE: 0xC001800C 0x00000540
+WRITE: 0xC0018000 0x003F8F40
+WRITE: 0xC0013210 0x0014B014A
+; Switch all clocks to back dividers
+WRITE: 0xC0013010 0x00009FFF
+WRITE: 0xC0018010 0x000007AA
diff --git a/tools/wtp/clocks-600-600.txt b/tools/wtp/clocks-600-600.txt
new file mode 100644
index 0000000..ccf944f
--- /dev/null
+++ b/tools/wtp/clocks-600-600.txt
@@ -0,0 +1,50 @@
+; Switch all clocks to REFCLOCK
+WRITE: 0XC0013010 0x00000000
+WRITE: 0XC0018010 0x00000000
+; TBG-A: SE vco_div 0x0,
+; DIFF vco_div 0x1, vco_range 0xd
+; tbg_N 0x48 KVCO = 2400 MHz
+WRITE: 0xC0013204 0x00C00091
+WRITE: 0xC0013204 0x00C00121
+WRITE: 0xC0013220 0x08030803
+WRITE: 0xC0013208 0x94011401
+WRITE: 0xC0013230 0x00020002
+WRITE: 0xC0013208 0x94011401
+WRITE: 0xC001320C 0x53E556E6
+WRITE: 0xC0013210 0x014A014D
+WRITE: 0xC001320C 0x53E556E6
+WRITE: 0xC0013204 0x00C00120
+WRITE: 0xC0013208 0x94011401
+WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
+DELAY: 0x00000100
+; TBG-B: SE vco_div 0x1,
+; DIFF vco_div 0x1, vco_range 0xb
+; tbg_N 0x3c KVCO = 2000 MHz
+WRITE: 0xC0013204 0x00C10120
+WRITE: 0xC0013204 0x00F10120
+WRITE: 0xC0013220 0x08030803
+WRITE: 0xC0013208 0x14019401
+WRITE: 0xC0013230 0x00020002
+WRITE: 0xC0013208 0x14019401
+WRITE: 0xC001320C 0x56E556E6
+WRITE: 0xC0013210 0x014B014D
+WRITE: 0xC001320C 0x56E656E6
+WRITE: 0xC0013204 0x00F00120
+WRITE: 0xC0013208 0x14019401
+WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
+DELAY: 0x00000100
+; Set clocks to 600/600 preset
+WRITE: 0xC0013014 0x00000000
+WRITE: 0xC0013004 0x2326202A
+WRITE: 0xC0013008 0x1A09AAA9
+WRITE: 0xC001300C 0x208B3482
+WRITE: 0xC0013000 0x0333C0FE
+WRITE: 0xC0018014 0x00180000
+WRITE: 0xC0018004 0x053154C8
+WRITE: 0xC0018008 0x00307880
+WRITE: 0xC001800C 0x00000940
+WRITE: 0xC0018000 0x003F8F40
+WRITE: 0xC0013210 0x0014B014D
+; Switch all clocks to back dividers
+WRITE: 0xC0013010 0x00009FFF
+WRITE: 0xC0018010 0x000007AA
diff --git a/tools/wtp/clocks-800-800.txt b/tools/wtp/clocks-800-800.txt
new file mode 100644
index 0000000..bcfc41b
--- /dev/null
+++ b/tools/wtp/clocks-800-800.txt
@@ -0,0 +1,50 @@
+; Switch all clocks to REFCLOCK
+WRITE: 0XC0013010 0x00000000
+WRITE: 0XC0018010 0x00000000
+; TBG-A: SE vco_div 0x0,
+; DIFF vco_div 0x1, vco_range 0xa
+; tbg_N 0x30 KVCO = 1600 MHz
+WRITE: 0xC0013204 0x00C00091
+WRITE: 0xC0013204 0x00C000C1
+WRITE: 0xC0013220 0x08030803
+WRITE: 0xC0013208 0x94011400
+WRITE: 0xC0013230 0x00020002
+WRITE: 0xC0013208 0x94011400
+WRITE: 0xC001320C 0x53E556E6
+WRITE: 0xC0013210 0x014A014A
+WRITE: 0xC001320C 0x53E556E5
+WRITE: 0xC0013204 0x00C000C0
+WRITE: 0xC0013208 0x94011400
+WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
+DELAY: 0x00000100
+; TBG-B: SE vco_div 0x1,
+; DIFF vco_div 0x2, vco_range 0xb
+; tbg_N 0x3c KVCO = 2000 MHz
+WRITE: 0xC0013204 0x00C100C0
+WRITE: 0xC0013204 0x00F100C0
+WRITE: 0xC0013220 0x08030803
+WRITE: 0xC0013208 0x14019400
+WRITE: 0xC0013230 0x00040002
+WRITE: 0xC0013208 0x14019400
+WRITE: 0xC001320C 0x56E556E5
+WRITE: 0xC0013210 0x014B014A
+WRITE: 0xC001320C 0x56E656E5
+WRITE: 0xC0013204 0x00F000C0
+WRITE: 0xC0013208 0x14019400
+WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
+DELAY: 0x00000100
+; Set clocks to 800/800 preset
+WRITE: 0xC0013014 0x00000000
+WRITE: 0xC0013004 0x1296202C
+WRITE: 0xC0013008 0x21061AA9
+WRITE: 0xC001300C 0x20543082
+WRITE: 0xC0013000 0x0003C0F2
+WRITE: 0xC0018014 0x00180000
+WRITE: 0xC0018004 0x02515508
+WRITE: 0xC0018008 0x00307880
+WRITE: 0xC001800C 0x00000540
+WRITE: 0xC0018000 0x003F8F40
+WRITE: 0xC0013210 0x0014B014A
+; Switch all clocks to back dividers
+WRITE: 0xC0013010 0x00009FFF
+WRITE: 0xC0018010 0x000007AA
diff --git a/tools/wtp/ddr-800.txt b/tools/wtp/ddr-800.txt
index 89cde60..c3225d2 100644
--- a/tools/wtp/ddr-800.txt
+++ b/tools/wtp/ddr-800.txt
@@ -1,53 +1,3 @@
-; Switch all clocks to REFCLOCK
-WRITE: 0XC0013010 0x00000000
-WRITE: 0XC0018010 0x00000000
-; TBG-A: SE vco_div 0x0,
-; DIFF vco_div 0x1, vco_range 0xa
-; tbg_N 0x30 KVCO = 1600 MHz
-WRITE: 0xC0013204 0x00C00091
-WRITE: 0xC0013204 0x00C000C1
-WRITE: 0xC0013220 0x08030803
-WRITE: 0xC0013208 0x94011400
-WRITE: 0xC0013230 0x00020002
-WRITE: 0xC0013208 0x94011400
-WRITE: 0xC001320C 0x53E556E6
-WRITE: 0xC0013210 0x014A014A
-WRITE: 0xC001320C 0x53E556E5
-WRITE: 0xC0013204 0x00C000C0
-WRITE: 0xC0013208 0x94011400
-WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
-DELAY: 0x00000100
-; TBG-B: SE vco_div 0x1,
-; DIFF vco_div 0x2, vco_range 0xb
-; tbg_N 0x3c KVCO = 2000 MHz
-WRITE: 0xC0013204 0x00C100C0
-WRITE: 0xC0013204 0x00F100C0
-WRITE: 0xC0013220 0x08030803
-WRITE: 0xC0013208 0x14019400
-WRITE: 0xC0013230 0x00040002
-WRITE: 0xC0013208 0x14019400
-WRITE: 0xC001320C 0x56E556E5
-WRITE: 0xC0013210 0x014B014A
-WRITE: 0xC001320C 0x56E656E5
-WRITE: 0xC0013204 0x00F000C0
-WRITE: 0xC0013208 0x14019400
-WAIT_FOR_BIT_SET: 0xC0013208 0x80008000 0x00100000
-DELAY: 0x00000100
-; Set clocks to 800/800 preset
-WRITE: 0xC0013014 0x00000000
-WRITE: 0xC0013004 0x1296202C
-WRITE: 0xC0013008 0x21061AA9
-WRITE: 0xC001300C 0x20543082
-WRITE: 0xC0013000 0x0003C0F2
-WRITE: 0xC0018014 0x00180000
-WRITE: 0xC0018004 0x02515508
-WRITE: 0xC0018008 0x00307880
-WRITE: 0xC001800C 0x00000540
-WRITE: 0xC0018000 0x003F8F40
-WRITE: 0xC0013210 0x0014B014A
-; Switch all clocks to back dividers
-WRITE: 0xC0013010 0x00009FFF
-WRITE: 0xC0018010 0x000007AA
 ; Set DDR to 800MHz
 ; Pre-init
 WRITE: 0xC0014008 0x00404500
diff --git a/tools/wtp/dll_tune.txt b/tools/wtp/dll_tune.txt
index 8e3cbb2..05b269b 100644
--- a/tools/wtp/dll_tune.txt
+++ b/tools/wtp/dll_tune.txt
@@ -1,3 +1,5 @@
+; DLL tune procedures - adjust read phase
+;
 BRANCH: DTUN				; Call mvebu_dram_dll_tune()
 ;
 ;************************************************
-- 
1.9.1

